src/coreclr/vm/greenthreads.cpp

```diff
@@ -401,20 +404,69 @@ SuspendedGreenThread* GreenThread_ResumeThread(SuspendedGreenThread* pSuspendedT
     ThreadBase* pOldThreadBase = GetThread()->GetActiveThreadBase();
     GetThread()->SetActiveThreadBase(pGreenThread);
 
+    t_greenThread.greenThreadOnStack = true;
+
     ResumeSuspendedThreadHelper();
+    t_greenThread.greenThreadOnStack = false;
 
     GetThread()->SetActiveThreadBase(pOldThreadBase);
     return ProduceSuspendedGreenThreadStruct(pGreenThread);
 }
 
 extern "C" void End_More_Thread_Bookeeping()
 {
-    if (t_greenThread.inGreenThread) // This should avoid doing this for the initial jump into the green thread and jumps from green to OS
+    if (t_greenThread.inGreenThread)
     {
         if (t_greenThread.pStackListCurrent->prev == NULL)
-            __debugbreak();
-        t_greenThread.pStackListCurrent = t_greenThread.pStackListCurrent->prev;
+        {
+            // We should only hit this path when a green thread is finishing
+            t_greenThread.inGreenThread = false;
+        }
```

Comment:
Do you want to keep the current pStackListCurrent value rather than null it out here?  (current = t = current->prev;  if (t == nullptr) inGT=false)?

Reply:
Its a bit ugly, but yes, I think I do. We use it as part of cleanup. :(